### DISCUSSION TITLE
How do I handle having dependencies on both 32 & 64 bit versions of a DLL when the file-name is the same for both?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#19404] [Q&A]
Our software has dozens of .EXE components which, with their DLL dependencies, get installed into a `BIN` folder on the user's PC. These are a mix of x86 & x64 binaries and in at least one case we have the same vcpkg package a dependency of x86 and x64 modules. So far we noticed it with `zlib1.dll` but there will likely be others.
vcpkg copies the required DLLs into our build directory but both 32/64 bit versions of zlib1.dll have the same name so they can't both exist with the same name.

Some packages are very good at naming DLLs - boost_atomic-vc142-mt-x64-1_76.dll is a great example - but others like zlib are not. Other than build everything with static linkage (really not ideal!) what can I do? Is it a bug? Could I force one or two packages to build statically somehow, while the rest still build to DLLs unless overridden?

**Comment by cenit** [comment]
> Could I force one or two packages to build statically somehow, while the rest still build to DLLs unless overridden?

yes, in the triplet you can set specific ports to be static (or dynamic in a static triplet).

good question yours also. Is the “single architecture” way impossible/not desired for you?

**Comment by jdx-john** [comment]
@cenit it is a large, very old legacy codebase so not really :) Certainly not within the scope of work I am doing.
I am wondering if ports should emit architecture-specific DLL file-names as a rule, or some are just kind enough to do so. If I find a package which doesn't observe this (such as zlib) what can I do?

**Comment by JackBoosY** [comment]
If you use install command, please modify _ports/zlib/portfile.cmake_ and add these code to `vcpkg_configure_cmake`:
```cmake
vcpkg_configure_cmake(
...
    OPTIONS_DEBUG
        -DSKIP_INSTALL_HEADERS=ON
        -DCMAKE_DEBUG_POSTFIX=${VCPKG_TARGET_ARCHITECTURE}
    OPTIONS_RELEASE
        -DCMAKE_RELEASE_POSTFIX=${VCPKG_TARGET_ARCHITECTURE}
)
```

Please note that only avaliable for zlib.

**Comment by dg0yt** [comment]
Please remember that Windows is just one target OS. 
I would propose to use separate directories for different architectures. This is what is normaly used on Linux, macOS, MinGW and Android.

> add these code to vcpkg_configure_cmake:
> ~~~
>       -DCMAKE_<CONFIG>_POSTFIX=

For each port modified like this, extra care must be taken to verify cmake wrappers, pc files,  and `...-config` scripts. It is hard enough to have them handle debug postfix right, and now I see a release postfix proposed. This is going to have a long tail of changes.

**Comment by jdx-john** [comment]
Thanks @JackBoosY / @dg0yt for the tip regarding `-DCMAKE_..._POSTFIX` I'll test that. Is this the sort of change that you might want pushing as a PR or is it for me locally? I don't know iof vcpkg has rules on lib/DLL naming that some projects are missing, or I'm a corner case and nobody else needs these changes.

**Comment by cenit** [comment]
The `CMAKE_<CONFIG>_POSTFIX` is explicitly unallowed on vcpkg. So the hint was intended as a suggestion for you to use, in an overlay port maybe of the official one

**Comment by JackBoosY** [comment]
> The `CMAKE_<CONFIG>_POSTFIX` is explicitly unallowed on vcpkg. So the hint was intended as a suggestion for you to use, in an overlay port maybe of the official one

Yeah, that's only a workaround.

**Comment by jdx-john** [comment]
As I understand it, zlib is built from older source code which might not even be maintained any more(?)
I don't really know what an overlay port entails, but if the upstream library cannot be modified is this realistic?

**Comment by JackBoosY** [comment]
@jdx-john  See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/specifications/ports-overlay.md).
